networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_mgmt.py

```diff
@@ -0,0 +1,824 @@
+                data = convertor(os_o, provider_o)
+                revision = meta.get(os_id).get("_revision")
+                if revision != None:
+                    data["_revision"] = revision
+                if resource_type == Provider.SG_MEMBERS:
+                    if self._sg_members_require_update(self.client.get(path).json(), data):
+                        o = self.client.put(path=path, data=data)
+                        return self.metadata_update(resource_type, o.json())
+                    else:
+                        return meta
+                else:
+                    o = self.client.put(path=path, data=data)
+                    return self.metadata_update(resource_type, o.json())
+        else:
+            if not delete:
+                LOG.info(report, "updated")
+                o = self.client.post(path=path, data=convertor(os_o, provider_o))
+                return self.metadata_update(resource_type, o.json())
+            LOG.info("Resource:%s with ID:%s already deleted.", resource_type, os_id)
+
+    def _metadata_refresh(self, provider):
+        meta = dict()
+        params = {"switching_profile_type": "QosSwitchingProfile"}\
+            if provider == API.PROFILES else dict()
+        
+        # NSX does not allow to filter by custom property
+        # Search API has hard limit of 50k objects (with cursor)
+        result = self.client.get_all(path=provider, params=params)
+
+        for o in result:
+            if o.get("_create_user") != "admin":
+                continue
+
+            tags = {t.get("scope"):t.get("tag") for t in o.get("tags", [])}
+
+            # TODO - enable for multiple Agents for a single NSX-T Manager
+            # if NSXV3_AGENT_SCOPE not in tags or \
+            #     tags[NSXV3_AGENT_SCOPE] != cfg.CONF.AGENT.agent_id:
+            #     continue
+
+            if provider == API.IPSETS:
+                cidrs = o.get("ip_addresses")
+                if len(cidrs) == 1 and "0.0.0.0" in cidrs[0]:
+                    # This is RULE specific IPSet
+                    continue
+            
+            # Set generation to most recent for NSGroups, always skip update
+            gen = int(time.time()) if provider == API.NSGROUPS \
+                else tags.get(NSXV3_GENERATION_SCOPE)
+
+            # TODO - check for collisions (Ambiguously) and perform actions
+            meta[o.get("display_name")] = {
+                "id": o.get("id"), 
+                "rev": tags.get(NSXV3_REVISION_SCOPE), # empty set for NSGroup
+                "gen": gen,
+                "_revision": o.get("_revision")
+            }
+        return meta
+
+    def metadata_refresh(self, resource_type):
+        if resource_type == Provider.SG_RULE:
+            return # Not cached
+        
+        resources = self._cache.get(resource_type)
+        backup = dict()
+        resources["backup"] = backup
+        eventlet.sleep(self._chache_refresh_window)
```

Comment:
Wouldn't this cause a delay of 30 seconds for every call to metadata_refresh (which is called by outdated())? Not sure this is intended instead - for example to just do the refresh only every 30 seconds - else return.

Reply:
Outdated => metadata_refresh for a resource type is called once on every synchronization. As the metadata is rebuilt by querying NSX-T multiple times to get the full state, there is a change to miss some ongoing create requests that might lead to ambiguosity. This timeout acts as a transaction log interval in which all active jobs must be completed.
The value of 30 seconds can be adjusted based on the size of NSX-T managers and the expected load.

Reply:
I think, the point is more that there is a `sleep` inside the function, but if you want do it timed, it would be more sensible to do that on the caller side of the function.
This way I might call this function in an RPC call, and that RPC call will timeout on the RPC-timeout, because it is potentially lower than  the value `_chache_refresh_window`.

Reply:
I see your point and it makes sense. I do not expect this call to be ever made in RPC context.
I would prefer to leave it that way until there is a need.